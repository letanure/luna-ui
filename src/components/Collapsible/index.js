import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';

/**
 * Collapsible component.
 * E.g.:
 * ```html
 * <Collapsible collapsed text="Labels">
 *  <ul>
 *    <li>Element 1.1</li>
 *  </ul>
 * </Collapsible>
 * ```
 */
class CollapsibleComponent extends React.Component {
  constructor(props) {
    super(props);
    const { collapsed } = props && props;
    this.state = { isCollapsed: collapsed };
  }

  _toggle(event) {
    if (event) {
      event.preventDefault();
    }

    this.setState((prevState) => ({
      isCollapsed: !prevState.isCollapsed
    }));
  }

  render() {
    const { text, className } = this.props;

    const CtxFontAwesomeIconRight = styled(FontAwesomeIcon)`
      right: 24px;
      position: absolute;
      ${(props) => `color: ${props.theme.collapsible.toggleIcon}`};
    `;

    const CollapsibleArea = styled.div`
      padding: 10px;
      ${this.state.isCollapsed ? 'display: none' : 'display: block'};
    `;

    const CollapsibleButton = styled.button`
      width: 100%;
      height: 50px;
      border: none;
      outline: none;
      cursor: pointer;
      font-size: 14px;
      text-align: left;
      font-weight: bold;
      position: relative;
      padding: 0 0 0 10px;
      transition: all 0.25s linear;
      color: ${(props) => props.theme.collapsible.label};

      background-color:
        ${(props) => this.state.isCollapsed ? props.theme.collapsible.inactiveBg : props.theme.collapsible.activeBg};

      &:hover {
        background-color: ${(props) => props.theme.collapsible.activeBg};
      }
    `;

    const toggleIcon = this.state.isCollapsed ? 'caret-down' : 'caret-up';

    return (
      <div className={className}>
        <CollapsibleButton onClick={this._toggle.bind(this)}>
          {text} <CtxFontAwesomeIconRight icon={toggleIcon} fixedWidth />
        </CollapsibleButton>
        <CollapsibleArea>{this.props.children}</CollapsibleArea>
      </div>
    );
  }
}

CollapsibleComponent.propTypes = {
  text: PropTypes.string,
  className: PropTypes.string,
  children: PropTypes.node,
  collapsed: PropTypes.bool
};

CollapsibleComponent.defaultProps = {
  text: '',
  className: '',
  children: null,
  collapsed: false
};

export default CollapsibleComponent;
