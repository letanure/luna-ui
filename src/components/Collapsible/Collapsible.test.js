import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import theme from '../../config/defaultTheme';
import ThemeProvider from '../Theme/index';

import Collapsible from './index';

describe('<Collapsible />', () => {
  let content;
  let component;
  let componentCollapsed;

  beforeEach(() => {
    content =
      <ul>
        <li>First item</li>
        <li>Second item</li>
        <li>Third item</li>
      </ul>
    ;

    component =
      <ThemeProvider theme={theme}>
        <Collapsible text="Labels">
          {content}
        </Collapsible>
      </ThemeProvider>;

    componentCollapsed =
      <ThemeProvider theme={theme}>
        <Collapsible collapsed text="Labels">
          {content}
        </Collapsible>
      </ThemeProvider>;
  });

  test('HandleClick', () => {
    const wrapper = mount(component);
    wrapper.find('button').simulate('click');
    const componentInstance = wrapper
      .find(Collapsible)
      .instance();
    expect(componentInstance.state.isCollapsed).toBeTruthy();
  });

  test('toMatchSnapshot: Expanded', () => {
    const tree = renderer
      .create(component)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('toMatchSnapshot: Collapsed', () => {
    const tree = renderer
      .create(componentCollapsed)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
