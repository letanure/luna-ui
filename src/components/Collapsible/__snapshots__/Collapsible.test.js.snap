// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Collapsible /> toMatchSnapshot: Collapsed 1`] = `
<div>
  <div
    className=""
  >
    <button
      className="sc-gZMcBi eAgGZT"
      onClick={[Function]}
    >
      Labels
       
    </button>
    <div
      className="sc-iwsKbI jLkQCk"
    >
      <ul>
        <li>
          First item
        </li>
        <li>
          Second item
        </li>
        <li>
          Third item
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<Collapsible /> toMatchSnapshot: Expanded 1`] = `
<div>
  <div
    className=""
  >
    <button
      className="sc-htoDjs bdCNyL"
      onClick={[Function]}
    >
      Labels
       
    </button>
    <div
      className="sc-gzVnrw kqQGtl"
    >
      <ul>
        <li>
          First item
        </li>
        <li>
          Second item
        </li>
        <li>
          Third item
        </li>
      </ul>
    </div>
  </div>
</div>
`;
